# ENTREGABLE 4

# INSTRUCCIONES

Utilizar el archivo CSV (`dataset_banco_clean.csv`) con 45189 filas y 17 columnas y aplicar las técnicas de normalización del entregable 3.

In [9]:
# imports
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd

In [10]:
ruta = "/content/dataset_banco_clean.csv"
df = pd.read_csv(ruta)

# Objetivo

Generar un model de clasificación capaz de predecir la clase de flor en función de las carácterísticas del dataset

* Aplicar las técnicas oportunas de procesamiento de datos

* Generar split de los datos

* Valorar diferentes modelos de clasificación

* Comparación entre modelos

* Ensemble

* Métricas

* Conclusiones finales

In [11]:
print(df.shape)
df.head()

(45189, 17)


age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no   2143.0     yes   no   
1   44    technician   single  secondary      no     29.0     yes   no   
2   33  entrepreneur  married  secondary      no      2.0     yes  yes   
3   47   blue-collar  married    unknown      no   1506.0     yes   no   
4   33       unknown   single    unknown      no      1.0      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may     261.0         1   -1.0         0  unknown  no  
1  unknown    5   may     151.0         1   -1.0         0  unknown  no  
2  unknown    5   may      76.0         1   -1.0         0  unknown  no  
3  unknown    5   may      92.0         1   -1.0         0  unknown  no  
4  unknown    5   may     198.0         1   -1.0         0  unknown  no

In [12]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Suponiendo que 'df' es tu conjunto de datos original
# y has definido las variables 'X' (características) e 'y' (objetivo) como en tu código anterior
X = df.drop(['y'], axis=1)
y = df['y']
# Convertir variables categóricas usando codificación one-hot
X_encoded = pd.get_dummies(X, drop_first=True)

# Dividir el conjunto de datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

# Inicializar y entrenar el modelo de Regresión Logística
model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = model.predict(X_test)

# Evaluar el rendimiento del modelo
print("Accuracy Score:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Accuracy Score: 0.8889134764328391

Confusion Matrix:
 [[7823  155]
 [ 849  211]]

Classification Report:
               precision    recall  f1-score   support

          no       0.90      0.98      0.94      7978
         yes       0.58      0.20      0.30      1060

    accuracy                           0.89      9038
   macro avg       0.74      0.59      0.62      9038
weighted avg       0.86      0.89      0.86      9038



In [13]:
from sklearn.tree import DecisionTreeClassifier

# Inicializar y entrenar el modelo de Árbol de Decisión
tree_model = DecisionTreeClassifier(random_state=42)
tree_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
tree_y_pred = tree_model.predict(X_test)

# Evaluar el rendimiento del modelo de Árbol de Decisión
print("Accuracy Score (Decision Tree):", accuracy_score(y_test, tree_y_pred))
print("\nConfusion Matrix (Decision Tree):\n", confusion_matrix(y_test, tree_y_pred))
print("\nClassification Report (Decision Tree):\n", classification_report(y_test, tree_y_pred))


Accuracy Score (Decision Tree): 0.8739765434830715

Confusion Matrix (Decision Tree):
 [[7419  559]
 [ 580  480]]

Classification Report (Decision Tree):
               precision    recall  f1-score   support

          no       0.93      0.93      0.93      7978
         yes       0.46      0.45      0.46      1060

    accuracy                           0.87      9038
   macro avg       0.69      0.69      0.69      9038
weighted avg       0.87      0.87      0.87      9038



In [14]:
from sklearn.svm import SVC

# Inicializar y entrenar el modelo de Máquinas de Soporte Vectorial
svm_model = SVC(random_state=42)
svm_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_svm_pred = svm_model.predict(X_test)

# Evaluar el rendimiento del modelo
print("Accuracy Score (SVM):", accuracy_score(y_test, y_svm_pred))
print("\nConfusion Matrix (SVM):\n", confusion_matrix(y_test, y_svm_pred))
print("\nClassification Report (SVM):\n", classification_report(y_test, y_svm_pred))


Accuracy Score (SVM): 0.8822748395662757

Confusion Matrix (SVM):
 [[7968   10]
 [1054    6]]

Classification Report (SVM):
               precision    recall  f1-score   support

          no       0.88      1.00      0.94      7978
         yes       0.38      0.01      0.01      1060

    accuracy                           0.88      9038
   macro avg       0.63      0.50      0.47      9038
weighted avg       0.82      0.88      0.83      9038



In [15]:
from sklearn.preprocessing import LabelEncoder
from sklearn.utils import class_weight
from xgboost import XGBClassifier
import numpy as np


# Inicializar el codificador de etiquetas
label_encoder = LabelEncoder()

# Convertir las clases 'no' y 'yes' a enteros
y_train_encoded = label_encoder.fit_transform(y_train)
y_test_encoded = label_encoder.transform(y_test)

# Obtener las clases únicas
unique_classes = np.unique(y_train)

# Calcular los pesos de clase
class_weights = class_weight.compute_class_weight('balanced', classes=unique_classes, y=y_train)

# Inicializar y entrenar el modelo de XGBoost con pesos de clase
xgb_model = XGBClassifier(random_state=42, scale_pos_weight=class_weights[1])
xgb_model.fit(X_train, y_train_encoded)

# Realizar predicciones en el conjunto de prueba
y_xgb_pred = xgb_model.predict(X_test)

# Evaluar el rendimiento del modelo
print("Accuracy Score (XGBoost):", accuracy_score(y_test_encoded, y_xgb_pred))
print("\nConfusion Matrix (XGBoost):\n", confusion_matrix(y_test_encoded, y_xgb_pred))
print("\nClassification Report (XGBoost):\n", classification_report(y_test_encoded, y_xgb_pred))


Accuracy Score (XGBoost): 0.8951095375082982

Confusion Matrix (XGBoost):
 [[7290  688]
 [ 260  800]]

Classification Report (XGBoost):
               precision    recall  f1-score   support

           0       0.97      0.91      0.94      7978
           1       0.54      0.75      0.63      1060

    accuracy                           0.90      9038
   macro avg       0.75      0.83      0.78      9038
weighted avg       0.92      0.90      0.90      9038



In [16]:
# Obtener la importancia de las características
feature_importances = xgb_model.feature_importances_

# Mostrar la importancia de las características
feature_importance_df = pd.DataFrame({'Feature': X_train.columns, 'Importance': feature_importances})
feature_importance_df = feature_importance_df.sort_values(by='Importance', ascending=False)
print(feature_importance_df)


                Feature  Importance
40     poutcome_success    0.194050
27      contact_unknown    0.136891
34            month_mar    0.056780
3              duration    0.048627
24          housing_yes    0.048585
41     poutcome_unknown    0.040193
37            month_oct    0.039505
33            month_jun    0.034584
30            month_feb    0.030171
38            month_sep    0.027384
32            month_jul    0.026879
31            month_jan    0.024996
28            month_aug    0.021859
35            month_may    0.020942
29            month_dec    0.018986
25             loan_yes    0.017984
36            month_nov    0.017265
5                 pdays    0.012760
2                   day    0.012264
7       job_blue-collar    0.009964
18      marital_married    0.009895
21   education_tertiary    0.009512
0                   age    0.009506
14          job_student    0.008939
4              campaign    0.008851
9         job_housemaid    0.008445
6              previous    0

In [17]:
from sklearn.model_selection import GridSearchCV
from xgboost import XGBClassifier

# Definir el espacio de búsqueda de hiperparámetros
param_grid = {
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 4, 5],
    'n_estimators': [50, 100, 200],
    'subsample': [0.8, 0.9, 1.0],
    'colsample_bytree': [0.8, 0.9, 1.0],
}

# Inicializar el modelo de XGBoost
xgb_model = XGBClassifier(random_state=42, scale_pos_weight=class_weights[1])

# Inicializar la búsqueda de cuadrícula
grid_search = GridSearchCV(xgb_model, param_grid, cv=3, scoring='accuracy')
grid_search.fit(X_train, y_train_encoded)

# Obtener los mejores hiperparámetros
best_params = grid_search.best_params_

# Inicializar y entrenar el modelo de XGBoost con los mejores hiperparámetros
best_xgb_model = XGBClassifier(random_state=42, scale_pos_weight=class_weights[1], **best_params)
best_xgb_model.fit(X_train, y_train_encoded)

# Realizar predicciones en el conjunto de prueba
y_xgb_pred = best_xgb_model.predict(X_test)

# Evaluar el rendimiento del modelo ajustado
print("Accuracy Score (XGBoost - Ajustado):", accuracy_score(y_test_encoded, y_xgb_pred))
print("\nConfusion Matrix (XGBoost - Ajustado):\n", confusion_matrix(y_test_encoded, y_xgb_pred))
print("\nClassification Report (XGBoost - Ajustado):\n", classification_report(y_test_encoded, y_xgb_pred))


Accuracy Score (XGBoost - Ajustado): 0.9046249170170392

Confusion Matrix (XGBoost - Ajustado):
 [[7824  154]
 [ 708  352]]

Classification Report (XGBoost - Ajustado):
               precision    recall  f1-score   support

           0       0.92      0.98      0.95      7978
           1       0.70      0.33      0.45      1060

    accuracy                           0.90      9038
   macro avg       0.81      0.66      0.70      9038
weighted avg       0.89      0.90      0.89      9038



In [18]:
from sklearn.ensemble import RandomForestClassifier

# Inicializar y entrenar el modelo de Random Forest
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train_encoded)

# Realizar predicciones en el conjunto de prueba
y_rf_pred = rf_model.predict(X_test)

# Evaluar el rendimiento del modelo de Random Forest
print("Accuracy Score (Random Forest):", accuracy_score(y_test_encoded, y_rf_pred))
print("\nConfusion Matrix (Random Forest):\n", confusion_matrix(y_test_encoded, y_rf_pred))
print("\nClassification Report (Random Forest):\n", classification_report(y_test_encoded, y_rf_pred))


Accuracy Score (Random Forest): 0.9049568488603673

Confusion Matrix (Random Forest):
 [[7775  203]
 [ 656  404]]

Classification Report (Random Forest):
               precision    recall  f1-score   support

           0       0.92      0.97      0.95      7978
           1       0.67      0.38      0.48      1060

    accuracy                           0.90      9038
   macro avg       0.79      0.68      0.72      9038
weighted avg       0.89      0.90      0.89      9038



In [19]:
from sklearn.metrics import f1_score

# Calcular el F1-score con promedio micro
f1_micro = f1_score(y_test_encoded, y_rf_pred, average='micro')

print("F1 Score (micro) for Random Forest:", f1_micro)


F1 Score (micro) for Random Forest: 0.9049568488603673


In [21]:
from sklearn.model_selection import GridSearchCV

# Definir el espacio de búsqueda de hiperparámetros
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['auto', 'sqrt', 'log2']
}

# Inicializar el modelo de Random Forest
rf_model = RandomForestClassifier(random_state=42)

# Inicializar la búsqueda de cuadrícula
grid_search_rf = GridSearchCV(rf_model, param_grid, cv=3, scoring='accuracy')
grid_search_rf.fit(X_train, y_train_encoded)

# Obtener los mejores hiperparámetros
best_params_rf = grid_search_rf.best_params_

# Inicializar y entrenar el modelo de Random Forest con los mejores hiperparámetros
best_rf_model = RandomForestClassifier(random_state=42, **best_params_rf)
best_rf_model.fit(X_train, y_train_encoded)


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomFor

RandomForestClassifier(max_depth=20, max_features='auto', min_samples_split=5,
                       n_estimators=200, random_state=42)

In [22]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Asumiendo que X_train, y_train_encoded, X_test, y_test_encoded están definidos en otra parte

# Inicializar y entrenar el modelo de Random Forest con los parámetros específicos
rf_model = RandomForestClassifier(max_features='auto', min_samples_split=10, n_estimators=200, random_state=42)
rf_model.fit(X_train, y_train_encoded)

# Realizar predicciones en el conjunto de prueba
y_rf_pred = rf_model.predict(X_test)

# Evaluar el rendimiento del modelo de Random Forest
print("Accuracy Score (Random Forest):", accuracy_score(y_test_encoded, y_rf_pred))
print("\nConfusion Matrix (Random Forest):\n", confusion_matrix(y_test_encoded, y_rf_pred))
print("\nClassification Report (Random Forest):\n", classification_report(y_test_encoded, y_rf_pred))


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Accuracy Score (Random Forest): 0.9092719628236335

Confusion Matrix (Random Forest):
 [[7805  173]
 [ 647  413]]

Classification Report (Random Forest):
               precision    recall  f1-score   support

           0       0.92      0.98      0.95      7978
           1       0.70      0.39      0.50      1060

    accuracy                           0.91      9038
   macro avg       0.81      0.68      0.73      9038
weighted avg       0.90      0.91      0.90      9038



In [23]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

# Suponiendo que `df` es tu DataFrame.

# Separar características y etiqueta
X = df.drop('y', axis=1)
y = df['y'].map({'yes': 1, 'no': 0})  # Codificar la etiqueta como 0 y 1

# Dividir el conjunto de datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Preprocesamiento: Codificar variables categóricas y escalar las numéricas
categorical_cols = [cname for cname in X_train.columns if X_train[cname].dtype == "object"]
numerical_cols = [cname for cname in X_train.columns if X_train[cname].dtype in ['int64', 'float64']]

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ])

# Modelos base
base_models = [
    ('lr', make_pipeline(preprocessor, LogisticRegression(max_iter=1000))),
    ('dt', make_pipeline(preprocessor, DecisionTreeClassifier())),
    ('svc', make_pipeline(preprocessor, SVC(probability=True)))
]

# Modelo meta
meta_model = LogisticRegression()

# Ensamble de Stacking
stacking_clf = StackingClassifier(estimators=base_models, final_estimator=meta_model)

# Entrenar el modelo ensamblado
stacking_clf.fit(X_train, y_train)

# Evaluar el modelo
print(f'Accuracy on test set: {stacking_clf.score(X_test, y_test):.2f}')


Accuracy on test set: 0.91


# Conclusiones

### Conclusiones:

1. **La precisión alcanzada es notable:** El modelo logró una precisión del 91% en el conjunto de prueba, lo cual sugiere que el ensamble de modelos utilizado está funcionando bien en general.

2. **La diversidad de modelos ha sido beneficiosa:** La inclusión de tres modelos base diferentes (regresión logística, árbol de decisiones y SVM) proporcionó diversidad al ensamble, lo que podría haber contribuido a un rendimiento robusto y a una mejor generalización del modelo.

3. **Se evidencia la importancia del preprocesamiento:** El preprocesamiento robusto, que incluyó la estandarización de características numéricas y la codificación one-hot de características categóricas, fue crucial para asegurar que los modelos funcionaran bien con diferentes tipos de datos.

### Posibles mejoras:

1. **Explorar más a fondo los hiperparámetros:** Aunque se realizó un GridSearch de parámetros y modelos, hay margen para una mayor exploración. Sería beneficioso buscar una combinación más amplia de hiperparámetros o probar modelos adicionales para mejorar aún más el rendimiento.

2. **Considerar la exploración exhaustiva de características:** La ingeniería de características puede marcar una gran diferencia en el rendimiento del modelo. Explorar diferentes formas de transformar o seleccionar características podría mejorar la precisión del modelo.

3. **Abordar desequilibrios en las clases de datos:** Si hay desequilibrios en las clases del conjunto de datos, sería útil explorar técnicas de muestreo o ajuste de pesos para mejorar la capacidad del modelo para detectar la clase minoritaria.

4. **Explorar otros tipos de ensembles:** Además del Stacking, sería interesante explorar otros tipos de ensembles, como Bagging o Boosting, para ver si ofrecen mejoras adicionales en el rendimiento.

5. **Realizar un análisis detallado de los errores:** Un análisis detallado de los errores cometidos por el modelo podría revelar patrones o tendencias específicas que podrían abordarse mediante ajustes en el modelo o en los datos.

En general, el modelo construido y el rendimiento alcanzado son satisfactorios. La clave ahora es seguir iterando y explorando diferentes estrategias para mejorar aún más su rendimiento.
